tests/conftest.py

```diff
@@ -2,29 +2,288 @@
+    k8s_storage_dependency.set_state(
+        k8s_client=MockK8sStorageClient(test_obj=obj_factory),
+        logger=logger,
+    )
+    return k8s_storage_dependency.k8s_client
+
+
+@pytest_asyncio.fixture
+async def gafaelfawr_storage_client(
+    http_client: AsyncClient,
+    obj_factory: TestObjectFactory,
+) -> GafaelfawrStorageClient:
+    gafaelfawr_storage_dependency.set_state(
+        client=MockGafaelfawrStorageClient(
+            token="", http_client=http_client, test_obj=obj_factory
+        )
+    )
+    return gafaelfawr_storage_dependency.gafaelfawr_client
+
+
+@pytest_asyncio.fixture
+async def docker_storage_client(
+    obj_factory: TestObjectFactory,
+    config: Configuration,
+    docker_credentials: DockerCredentialsMap,
+    logger: BoundLogger,
+    http_client: AsyncClient,
+) -> DockerStorageClient:
+    docker_storage_dependency.set_state(
+        docker_client=MockDockerStorageClient(test_obj=obj_factory),
+        logger=logger,
+        http_client=http_client,
+        config=config,
+    )
+    return docker_storage_dependency.docker_client
+
+
+@pytest_asyncio.fixture
+async def prepuller_manager(
+    config: Configuration,
+    k8s_storage_client: K8sStorageClient,
+    docker_storage_client: DockerStorageClient,
+    logger: BoundLogger,
+) -> PrepullerManager:
+    prepuller_manager_dependency.set_state(
+        logger=logger,
+        config=config,
+        k8s_client=k8s_storage_client,
+        docker_client=docker_storage_client,
+    )
+    return prepuller_manager_dependency.prepuller_manager
+
+
+@pytest_asyncio.fixture
+async def app(
+    k8s_storage_client: K8sStorageClient,
+    docker_storage_client: DockerStorageClient,
+    gafaelfawr_storage_client: GafaelfawrStorageClient,
+) -> AsyncIterator[FastAPI]:
     """Return a configured test application.
 
     Wraps the application in a lifespan manager so that startup and shutdown
     events are sent during test execution.
     """
-    async with LifespanManager(main.app):
-        yield main.app
+    app = create_app(
+        config_dir=str(CONFIG_DIR),
+        storage_clients=StorageClientBundle(
+            k8s_client=k8s_storage_client,
+            docker_client=docker_storage_client,
+            gafaelfawr_client=gafaelfawr_storage_client,
+        ),
+    )
+    async with LifespanManager(app):
+        yield app
 
 
 @pytest_asyncio.fixture
-async def client(app: FastAPI) -> AsyncIterator[AsyncClient]:
+async def app_client(
+    app: FastAPI, config: Configuration
+) -> AsyncIterator[AsyncClient]:
     """Return an ``httpx.AsyncClient`` configured to talk to the test app."""
-    async with AsyncClient(app=app, base_url="https://example.com/") as client:
+    async with AsyncClient(
+        app=app, base_url=config.runtime.instance_url
+    ) as client:
+        yield client
+
+
+@pytest_asyncio.fixture
+async def user_map(obj_factory: TestObjectFactory) -> UserMap:
+    return obj_factory.usermap
```

Comment:
I'm not sure it makes sense for this to be a fixture. I assume the idea is to test queries when user's lab pods are in various states of spawning, but I think it would be better to start from an idle controller and then spawn a lab pod, followed by asking the API for information at various stages in that process.

Reply:
removed